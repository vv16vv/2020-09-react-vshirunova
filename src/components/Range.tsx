import React, {ChangeEvent, useCallback, useState} from "react";
import {RangeBlock, StyledRange} from "./StyledComponents";
import {CenteredLabel, RangeLabel} from "./StyledTextComponents";

export interface RangeProp {
    name: string;
    min?: number;
    max: number;
    step?: number;
    value: number;
    changeHandler: (newValue: number) => void;
}

export const Range: React.FC<RangeProp> = ({name, min, max, step, value, changeHandler}) => {
    const realMin = min ?? 1;
    const [currentValue, setCurValue] = useState(value);

    const onChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
        const newValue = +event?.target?.value ?? realMin;
        changeHandler(newValue);
        setCurValue(newValue);
    }, []);

    return (<RangeBlock>
        <div><CenteredLabel>{currentValue}</CenteredLabel></div>
        <div>
            <RangeLabel>{realMin}</RangeLabel>
            <StyledRange
                name={name}
                type="range"
                min={realMin}
                max={max}
                step={step ?? 1}
                value={currentValue}
                onChange={onChange}
            />
            <RangeLabel>{max}</RangeLabel>
        </div>
    </RangeBlock>)
};