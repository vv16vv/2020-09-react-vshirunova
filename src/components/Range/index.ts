export {Range} from "./Range";
export {RangeFormik} from "./RangeFormik";