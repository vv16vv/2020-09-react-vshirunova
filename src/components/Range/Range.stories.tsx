import React from "react";
import {action} from "@storybook/addon-actions";
import {number, withKnobs} from "@storybook/addon-knobs";

import {Range} from "./Range"
import {Form, Formik} from "formik";

export default {
    title: "Range Story",
    decorators: [withKnobs],
}

export const FullRangeStory: React.FC<{}> = () => {
    const min = number('Min', 1);
    const max = number('Max', 10);
    const step = number('Step', 2);
    const value = number('Value', 4);
    return (
        <Formik initialValues={{test: value}} onSubmit={action("Form submitted")}>
            <Form>
                <Range
                    name={"test"}
                    min={min} max={max} step={step}
                    value={value}
                    changeHandler={action("Value changed")}/>
            </Form>
        </Formik>
    );
}

const PartRange: React.FC<{ max: number, initialValue: number }> = (props) => {
    const max = number('Max', props.max);
    const value = number('Value', props.initialValue);
    return (
        <Formik initialValues={{test: value}} onSubmit={action("Form submitted")}>
            <Form>
                <Range name={"test"}
                       max={max}
                       value={value}
                       changeHandler={action("Value changed")}/>
            </Form>
        </Formik>
    );
};

export const PartRangeStory: React.FC<{}> = () => {
    return <PartRange max={10} initialValue={4}/>
}
