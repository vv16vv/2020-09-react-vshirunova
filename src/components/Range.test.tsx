import React from "react";
import renderer from "react-test-renderer";
import {shallow} from "enzyme";
import {Range} from "./Range";

describe("Range", () => {

    const changeHandler = jest.fn();
    const min = 1;
    const max = 10;
    const step = 1;
    const initialValue = 5;

    it("should show min, max and current value - all props", () => {
        expect(renderer.create(
            <Range
                min={min}
                max={max}
                step={step}
                value={initialValue}
                changeHandler={changeHandler}/>
            ).toJSON()
        ).toMatchSnapshot();
    });

    it("should show min, max and current value - only required props", () => {
        expect(renderer.create(
            <Range
                max={max}
                value={initialValue}
                changeHandler={changeHandler}/>
            ).toJSON()
        ).toMatchSnapshot();
    });

    it("should set min and step when they missed", () => {
        const wrapper = shallow(
            <Range
                max={max}
                value={initialValue}
                changeHandler={changeHandler}/>
        )
            .find({type: 'range'})
            .props();
        expect(wrapper).toHaveProperty(['min'], 1)
        expect(wrapper).toHaveProperty(['step'], 1)
    });

    it('should call onChange handler', () => {
        const wrapper = shallow(
            <Range
                min={min}
                max={max}
                step={step}
                value={initialValue}
                changeHandler={changeHandler}/>
        );
        wrapper
            .find({type: 'range'})
            .simulate('change');

        expect(changeHandler).toBeCalled();
    });
});