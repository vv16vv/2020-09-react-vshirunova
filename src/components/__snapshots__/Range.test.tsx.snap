// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range should show min, max and current value - all props 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-814x24{width:200px;height:15px;background-color:#ffffff;}",
      }
    }
    data-emotion="css 814x24"
  />,
  <div
    className="css-814x24"
  >
    <div>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-1eqgw2z{font-family:Tahoma,serif;font-style:normal;font-weight:bold;font-size:15px;line-height:18px;text-align:center;width:100%;color:#00008b;background-color:#ffffff;}",
          }
        }
        data-emotion="css 1eqgw2z"
      />
      <legend
        className="css-1eqgw2z"
      >
        5
      </legend>
    </div>
    <div>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-1y5wkk4{font-family:Tahoma,serif;font-style:normal;font-weight:normal;font-size:13px;line-height:15px;color:#000000;display:inline;vertical-align:top;width:10%;color:#888888;background-color:#ffffff;}",
          }
        }
        data-emotion="css 1y5wkk4"
      />
      <span
        className="css-1y5wkk4"
      >
        1
      </span>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-oi2fxo{width:85%;color:#000000;background-color:#ffffff;}",
          }
        }
        data-emotion="css oi2fxo"
      />
      <input
        className="css-oi2fxo"
        max={10}
        min={1}
        onChange={[Function]}
        step={1}
        type="range"
        value={5}
      />
      <span
        className="css-1y5wkk4"
      >
        10
      </span>
    </div>
  </div>,
]
`;

exports[`Range should show min, max and current value - only required props 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-814x24{width:200px;height:15px;background-color:#ffffff;}",
      }
    }
    data-emotion="css 814x24"
  />,
  <div
    className="css-814x24"
  >
    <div>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-1eqgw2z{font-family:Tahoma,serif;font-style:normal;font-weight:bold;font-size:15px;line-height:18px;text-align:center;width:100%;color:#00008b;background-color:#ffffff;}",
          }
        }
        data-emotion="css 1eqgw2z"
      />
      <legend
        className="css-1eqgw2z"
      >
        5
      </legend>
    </div>
    <div>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-1y5wkk4{font-family:Tahoma,serif;font-style:normal;font-weight:normal;font-size:13px;line-height:15px;color:#000000;display:inline;vertical-align:top;width:10%;color:#888888;background-color:#ffffff;}",
          }
        }
        data-emotion="css 1y5wkk4"
      />
      <span
        className="css-1y5wkk4"
      >
        1
      </span>
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-oi2fxo{width:85%;color:#000000;background-color:#ffffff;}",
          }
        }
        data-emotion="css oi2fxo"
      />
      <input
        className="css-oi2fxo"
        max={10}
        min={1}
        onChange={[Function]}
        step={1}
        type="range"
        value={5}
      />
      <span
        className="css-1y5wkk4"
      >
        10
      </span>
    </div>
  </div>,
]
`;
